recipes/asyncpp/all/conanfile.py

```diff
@@ -0,0 +1,102 @@
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.build import check_min_cppstd
+from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
+from conan.tools.env import VirtualBuildEnv
+from conan.tools.files import copy, get, rm, rmdir
+from conan.tools.scm import Version
+import os
+
+
+required_conan_version = ">=1.53.0"
+
+
+class AsyncppRecipe(ConanFile):
+    name = "asyncpp"
+    description = "A C++20 coroutine library for asynchronous and parallel programming."
+    license = "MIT"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/project/package"
+    topics = ("coroutine", "c++20", "async", "parallel", "concurrency")
+
+    package_type = "library"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+    }
+
+    @property
+    def _min_cppstd(self):
+        return 20
+
+    @property
+    def _compilers_minimum_version(self):
+        return {
+            "gcc": "13",
+            "clang": "17",
+            "msvc": "193",
+            "Visual Studio": "17",
+        }
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            self.options.rm_safe("fPIC")
+
+    def layout(self):
+        # src_folder must use the same source folder name the project
+        cmake_layout(self, src_folder="src")
+
+    def validate(self):
+        # validate the minimum cpp standard supported. For C++ projects only
+        if self.settings.compiler.cppstd:
+            check_min_cppstd(self, self._min_cppstd)
+        minimum_version = self._compilers_minimum_version.get(str(self.settings.compiler), False)
+        if minimum_version and Version(self.settings.compiler.version) < minimum_version:
+            raise ConanInvalidConfiguration(
+                f"{self.ref} requires C++{self._min_cppstd}, which your compiler does not support."
+            )
+
+        if self.settings.compiler == "apple-clang":
+            # As per https://en.cppreference.com/w/cpp/compiler_support, apple-clang 14.0.0 partially supports C++20 but not jthread
+            raise ConanInvalidConfiguration(f"{self.ref} does not support apple-clang compiler, as it lacks jthread support.")
```

Comment:
```suggestion
            raise ConanInvalidConfiguration(f"{self.ref} does not support apple-clang compiler, as it lacks jthread support.")
            
        if is_msvc(self) and self.options.shared:
            # INFO: asyncpp does not contain .def neither provides .lib when building shared
            raise ConanInvalidConfiguration(f"{self.ref} does not support -o {self.ref}:shared=True with compiler=msvc")
```

The current error is because the project only provides a DLL file, but does not expose symbols in a .lib: https://c3i.jfrog.io/c3i/misc-v2/logs/pr/23974/8-windows-msvc/asyncpp/0.2.3//4dfc129aabdd7b56484b8f1e47f0337c1164e51e-build.txt

```
asyncpp/0.2.3: RUN: cmake --install "C:\J\workspace\prod-v2\bsr\39245\fbfcf\p\b\async8b8db22574e05\b\build" --config Release --prefix "C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p"
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/lib/asyncpp.dll
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/CMakeLists.txt
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/concepts.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/container
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/container/atomic_collection.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/container/atomic_deque.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/container/atomic_item.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/container/atomic_stack.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/event.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/generator.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/join.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/lock.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/memory
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/memory/rc_ptr.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/mutex.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/promise.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/scheduler.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/semaphore.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/shared_mutex.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/sleep.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/stream.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/task.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/testing
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/testing/interleaver.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/testing/suspension_point.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/threading
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/threading/cache.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/threading/spinlock.hpp
-- Installing: C:/J/workspace/prod-v2/bsr/39245/fbfcf/p/b/async8b8db22574e05/p/include/asyncpp/thread_pool.hpp
```

People would need to use WINAPI LoadLibrary and load symbols manually ... 

As you are the author, you could adjust the project to export symbols:

- Adding the CMake definition WINDOWS_EXPORT_ALL_SYMBOLS. It works most of cases, but is fragile, will expose all methods, including internals. It does not export static public variables, it will result in runtime errors
- Use __declspec to export only what you want: https://learn.microsoft.com/en-us/cpp/build/exporting-from-a-dll?view=msvc-170
- Use CMake GenerateExportHeader to automatically export functions: https://www.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/

Reply:
Hey @RubenRBS and @uilianries, thanks for checking out this issue. I won't have time in the upcoming week, but I'm going to modify the library so that the symbols are exported properly and get back to you.